### PR TITLE
add option to check a dns-cache instead of secondary nameservers

### PR DESCRIPTION
**Pull request by Farom**
Signed-off-by: Björn Lässig <b.laessig@pengutronix.de>

### CODE REVIEW

File: check_dnsserials [hunk 1]
```diff
@@ -147,8 +151,14 @@ if [ -z "$serial" ]; then
...
 
+if [ -n $cache ]; then
+        nameserverL="$cache"
+else
+        nameserverL=$($COMMAND_DIG +short NS @$master $domain)
+fi
+
 # iterate the NS-RRs grabbed from master dns-server for the domain
-for server in $($COMMAND_DIG +short NS @$master $domain); do
+for server in $nameserverL; do
```
Comment (bebehei) [inline check_dnsserials:161]: Could you please enlighten me, what the `L` stands for?

**Review by bebehei — COMMENTED** [review]
Interesting use of my script. But what's actually your goal? With a cache server (aka resolver), you can't actually check the specific authoritative servers for their specific serial.

Taking a wild guess, that you want to check a resolver for its cached serial. And I think `check_dig` does this much better:

```
/usr/lib/monitoring-plugins/check_dig -H <cache server> -l bebehei.de -T SOA -a "ns0.b3be.de. bebe.bebehei.de. 2018092604"
```

`check_dnsserials`'s use case is actually to monitor and check, that all authoritative name servers of a specific zone are in sync. How does this change contribute to it?

File: check_dnsserials (hunk 1 above)
Comment (Farom) [reply check_dnsserials:161]: It stands for List.

I am open for suggestions :-)

**Comment by Farom** [comment]
> Taking a wild guess, that you want to check a resolver for its cached serial.

I have caching servers, that get notifies, when a zone is changing. They pull the important zones via AXFR to cache the whole zone. When the zone is changing, they get a notify and use AXFR/IXFR again. So if my master on Internet is not reachable for various reasons, i have a full week of working DNS.

>  And I think check_dig does this much better:

I have to know the serial  which is in a permanent change so i cannot use check_dig.

> check_dnsserials's use case is actually to monitor and check,
> that all authoritative name servers of a specific zone are in sync.
> How does this change contribute to it?

There are lot of reasons to have non-authoritative nameservers, that cache a zone via AXFR/IXFR. You get strange errors, when zones are out of sync so you want to monitor them. With this change, check_dnsserial  can monitor and  check that non-authoritative name servers of a specific zone are in sync.

Maybe this would be some more obvious, if i would add  `+norecurse` in the line
`serial_slave=$($COMMAND_DIG +short SOA @$server $domain | cut -d " " -f 3) `
?

**Comment by bebehei** [comment]
What about instead of calling those servers "cache", calling them "slaves"? Any objections?

**Comment by bebehei** [comment]
So, I've taken the time to rewrite and harden the script. There had been so many unquoted values, it scared me.

Also I've added the ability to specify slave servers in 0d8bbd6aab0dbeb2d5a2f323f40d2fa123e10765. I'd appreciate if you check out the [`dev`](https://github.com/bebehei/check_dnsserials/compare/dev) branch and give it a try and feedback.

**Comment by Farom** [comment]
Looks good for me, will test my usecase today.

While working on hardening ... would you consider to add '+norecurse' to this line?
`${COMMAND_DIG}" +short +norecurse SOA "@${server}" "${domain}`
?

        +[no]recurse
           Toggle the setting of the RD (recursion desired) bit in the query.
           This bit is set by default, which means dig normally sends
           recursive queries. Recursion is automatically disabled when the
           +nssearch or +trace query options are used.

I want to check, whether the Slave can answer the request for SOA, not that it can answer it by questioning a primary/secondary NS.

**Comment by bebehei** [comment]
Yeah, I've seen your commit, but I didn't have the logical power to understand, what it'll actually affect. But yeah, it's a very good point, adding `+norecurse` here.

I cherry picked it onto dev and will merge it today on master.

**Comment by bebehei** [comment]
I've picked your commit and it's in here: 72e4861b1c2f4d6eb7a3f18f72e8cd6c00c19ed0

I've put it into a new release 1.1.0. It's on icingaexchange also published. Sorry for killing your first commit, but a big thank you for this PR!

**Comment by Farom** [comment]
Thanks for your time and work.

**Pull request closed without merging**